### PR TITLE
Add view_simple as meta function for view,  and avoid calling reshape_view_helper for unbacked

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #154757

address https://github.com/pytorch/pytorch/issues/153303
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Review by bobrenjc93 — APPROVED** [review]

**Comment by bobrenjc93** [comment]
In the future ideally make the change and have the code movement in a separate PR so it's easier for reviewers.

**Comment by laithsakka** [comment]
@sanchitintel I am changing the view meta, but I am running into some inductor CPU specific tests that are failing, can you review this PR. Can you tell why the new reshape is problematic for those test?

**Comment by laithsakka** [comment]
> @sanchitintel I am changing the view meta, but I am running into some inductor CPU specific tests that are failing, can you review this PR. Can you tell why the new reshape is problematic for those test?

here is an observation for this test before the change the output was
```
torch.Size([s38*s86, 384, 96]), (36864, 96, 1)
```
after it is
```
result is (((36864*s38*s86)//(s57*s87)), s57, s87), [s57*s87, s87, 1]
```
are those different though?

they are not different but the result is less specialized for the second case which is also correct.
```
note that s86 = 5, s87 = 96 , s57 =384, s38=8
```
hence doing substitution
we will get  
```
[s38*s86, 384, 96], [s57*s87, s87, 1] (36864, 96, 1) 
```
so result in fact are correct and same.


I do not think Inductor optimizations should depend  on Meta decomp specializations! , id Inductor want do an opt
let it specialize

**Comment by laithsakka** [comment]
I debugged those further more, and it seems that tuned_bmm is behaving differently because use_cpp_gemm_template
does not support dynamic n and k.

one way to fix this is :
1) to specialize n, k.
```
    # TODO(jgong5): support dynamic shapes for n or k
    if has_free_symbols((n, k)):
        return False
```
but i am unsure if we want to always do that. 
2) alternatively we can just  rewrite the test?
cc @frost-intel @sanchitintel

**Comment by sanchitintel** [comment]
Sorry, @laithsakka, I had missed the notifications :(
Please ping me on Slack if that happens again (although I'll try to ensure that it wouldn't).

@frost-intel, looks like some dynamic shape UTs pertaining to BMM are failing.
Does the BMM implementation depend upon some specific shape inference nuances for dynamic shape support? IIRC, it does.
Can you please take a look? Thanks!

**Comment by leslie-fang-intel** [comment]
Hi @laithsakka, since @frost-intel is taking PTO until Jun 17. I will take a look of this failure. Take `test_bmm_with_pointwise_with_reshape_dynamic_shapes_bs_5_Mdim_384_Kdim_96_Ndim_64_cpu_bfloat16` for example 

- The major difference after https://github.com/pytorch/pytorch/pull/154757, I see is that this case can't run it GEMM Template due to the check failure that https://github.com/pytorch/pytorch/blob/be16f21ca62ff04a88a11af330c4f7156ef1d158/torch/_inductor/utils.py#L1798-L1799 In the GEMM Template, we don't support the case when N or K is symbolic.

- For this test case, since the N, K has been marked as static through https://github.com/pytorch/pytorch/blob/be16f21ca62ff04a88a11af330c4f7156ef1d158/test/inductor/test_cpu_select_algorithm.py#L2692-L2695, do you know why it will become symbolic after this PR?

**Comment by laithsakka** [comment]
> Hi @laithsakka, since @frost-intel is taking PTO until Jun 17. I will take a look of this failure. Take `test_bmm_with_pointwise_with_reshape_dynamic_shapes_bs_5_Mdim_384_Kdim_96_Ndim_64_cpu_bfloat16` for example
> 
> * The major difference after [Add view_simple as meta function for view,  and avoid calling reshape_view_helper.  #154757](https://github.com/pytorch/pytorch/pull/154757), I see is that this case can't run it GEMM Template due to the check failure that https://github.com/pytorch/pytorch/blob/be16f21ca62ff04a88a11af330c4f7156ef1d158/torch/_inductor/utils.py#L1798-L1799
>    In the GEMM Template, we don't support the case when N or K is symbolic.
> * For this test case, since the N, K has been marked as static through https://github.com/pytorch/pytorch/blob/be16f21ca62ff04a88a11af330c4f7156ef1d158/test/inductor/test_cpu_select_algorithm.py#L2692-L2695
>   , do you know why it will become symbolic after this PR?

yes reshape_view_helper used to specialize unnecessarily.
and the new reshape does less specializations. 

consider this program for example, after  this PR  we only need one compilation for the program.

```
import torch
@torch.compile()
def func(x, y):
    return x.reshape(-1, y.size()[0], y.size()[1])

x = torch.rand(100,200, 10, 20)
torch._dynamo.mark_dynamic(x, 0)
torch._dynamo.mark_dynamic(x, 1)

y = torch.rand(20, 40)
torch._dynamo.maybe_mark_dynamic(y, 0)
torch._dynamo.maybe_mark_dynamic(y, 1)
# mark first two dims dynamic

func(x,y)
func(x,torch.rand(40, 20))
```
in the case of the unit test failing in this PR the following specializations are avoided. 
Screenshot 2025-06-05 at 11 03 12 AM

**Comment by laithsakka** [comment]
ah i see you want N, and K to be static .. let me look into this maybe they used to be dynamic and get specialized let me check this API you use to mark them static.  

Those are not dynamic because they are coming from the globals vars
    @parametrize("Kdim", (96,))
    @parametrize("Ndim", (64, 65))
  which are not marked dyanmic
  do you want me to mark those variables static. I think this test is probably using dynamic=True
  which is not needed since we mark things explicitly.

**Comment by sanchitintel** [comment]
> do you want me to mark those variables static

Yes, thanks!

> I think this test is probably using dynamic=True

It's not using `dynamic=True` in its `torch.compile` invocation

**Comment by laithsakka** [comment]
> > do you want me to mark those variables static
> 
> Yes, thanks!
> 
> > I think this test is probably using dynamic=True
> 
> It's not using `dynamic=True` in its `torch.compile` invocation

got it to froce them static we can do:
```
                result = x.reshape(-1, int(Mdim), int(Kdim)) @ other.reshape(-1, int(Kdim), int(Ndim))
 ```
 let me update PR 
 another thing to consider is something like this
 ```
   if not has_free_unbacked_symbols(n) and not has_free_unbacked_symbols(k):
       n = V.graph.sizevars.evaluate_static_shape(n)
       k = V.graph.sizevars.evaluate_static_shape(k)

    if has_free_symbols((n, k)):
        return False

....

    ```

**Comment by sanchitintel** [comment]
Thanks, @laithsakka!

The first change LGTM for passing UTs. EDIT: it now seems it's not passing UTs. I had wrongly assumed that the change was made to pass UTs.

However, we may need to consider something like the second change, so as to support templated BMM kernel without having to modify the source-code of models.

**Comment by laithsakka** [comment]
> Thanks, @laithsakka!
> 
> The first change LGTM for passing UTs.
> 
> However, we may need to consider something like the second change, so as to support templated BMM kernel without having to modify the source-code of models.

@bobrenjc93  we might want force static list like the dynamic allow list.

yeh @sanchitintel agree changing models code to force static not ideal unfortuelty we do not have mark_static for integers. right now.

@bobrenjc93  we might want force static list similar to the dynamic allow list.

**Comment by sanchitintel** [comment]
Thanks for the info, @laithsakka!

> we do not have mark_static for integers. right now

Can you please elaborate? We used `mark_static(tensor, dim)`, so that the corresponding shape could be marked static.

Please help clarify a doubt regarding the second change you recommended.

It seems (I might be wrong) that `evaluate_static_shape` may provide a shape hint for a symbol, if it's not constant.
If that's true, then if `k` and `n` would be different in subsequent invocations of a compiled model's forward, can we assume that recompilation would not occur? If the answer to this question is yes, then the templated BMM kernel that we'd have compiled earlier would crash or produce incorrect output for different `k` and/or `n` than the ones encountered at compile-time, since it'd only work for static `k` and `n`. (In practice, it won't be a problem in the first place if `k` and `n` dims of the BMMs in a model won't be different with inputs of different shapes to the model).

Also, to clarify, the failing UTs were not using `dynamic=True` in `torch.compile` invocations, but I should've mentioned earlier that they use the Dynamo config `dynamic_shapes` as `True`. Sorry about that, but it seems that config is True by default!

File: test/inductor/test_cpu_select_algorithm.py [hunk 1]
```diff
@@ -2680,7 +2680,10 @@ def __init__(self):
                 self.epilogue = torch.nn.ReLU()
 
             def forward(self, x, other, noise):
-                result = x.reshape(-1, Mdim, Kdim) @ other.reshape(-1, Kdim, Ndim)
+                # we do not want to to have Mdim, Kdim and Ndim dynamic so we force specialize them.
+                result = x.reshape(-1, int(Mdim), int(Kdim)) @ other.reshape(
+                    -1, int(Kdim), int(Ndim)
+                )
```
Comment (sanchitintel) [inline test/inductor/test_cpu_select_algorithm.py]: Sorry, @laithsakka, I had assumed that this change had worked at your end, so I mentioned that it LGTM. I'm a bit confused about it, though. It seems they were already integers.

Also, https://docs.pytorch.org/docs/stable/torch.compiler_dynamic_shapes.html#abridged-public-api mentions `torch._dynamo.mark_dynamic`. Likewise, we also used `torch._dynamo.mark_static` in the UTs.
This PR seems to override that behavior. Is that expected? 

It seems the intention of some of these UTs was to emulate the case in which batch-size may change for BMM at runtime.
After this PR, with real workloads, we may not be able to use templated BMM kernels on x86_64 machines in those scenarios. We could use something similar to the second change that you mentioned to enable its use, but when `k` and `n` of a BMM could have different values at runtime (than at compile-time), then the workload _may_ crash at run-time, if (depending upon the model) `k` & `n` would be different at runtime (that may not be the case for all models).

Comment (sanchitintel) [reply test/inductor/test_cpu_select_algorithm.py]: Just noticed that `TestDynamicShapes.test_view_copy_symint_with_static_input_dyn_input_shape2` had also been failing earlier in the CI job `linux-focal-py3_9-clang9-xla`.

**Comment by sanchitintel** [comment]
> reshape_view_helper used to specialize unnecessarily.
> and the new reshape does less specializations.

But is it supposed to override `torch._dynamo.mark_static`?
Sorry, does `torch._dynamo.mark_static` mean that the corresponding dim shape should _always_ be specialized (so that the traced shape should always be used)? It isn't being used much in the codebase, so I'm not sure.

File: test/inductor/test_cpu_select_algorithm.py [hunk 2]
```diff
@@ -2674,16 +2674,22 @@ def forward(self, x, other):
...
         class M(torch.nn.Module):
             def __init__(self):
                 super().__init__()
                 self.epilogue = torch.nn.ReLU()
 
             def forward(self, x, other, noise):
-                result = x.reshape(-1, Mdim, Kdim) @ other.reshape(-1, Kdim, Ndim)
+                # we do not want to to have Mdim, Kdim and Ndim dynamic so we force specialize them.
+                result = x.reshape(
+                    -1, guard_scalar(Mdim), guard_scalar(Kdim)
```
Comment (leslie-fang-intel) [inline test/inductor/test_cpu_select_algorithm.py]: Thanks. Is the model code change like `guard_scalar` a must have for this case? I am wondering this change may cause some real workload failed run into BMM Template if `guard_scalar` is a must.

File: test/inductor/test_cpu_select_algorithm.py (hunk 1 above)
Comment (laithsakka) [reply test/inductor/test_cpu_select_algorithm.py]: @sanchitintel 
1) yes int(x) did not specialize as intended, i changed to support guard_scalar in user code and used guard_scalar(x) instead to force specializations. 
2) mark_static should work find, mark_static only works on tensor dimensions, in this case Mdim, Kdim, Ndim are integers inputs to the graph. Those are not marke_dynamic nor static but left for torch.compile .
The PR does not change mark_dynamic, mark_static behaviour. They used to be dynamic before the PR and used to be specialized during the reshape. but the not anymore wit the PR.

**Comment by laithsakka** [comment]
> evaluate_static_shape

evaluate_static_shape will guard on the hint  meaning that a recompilation will happen if those change
```
    def evaluate_static_shape(self, left: Union[Expr, int]) -> int:
        if isinstance(left, int):
            return left
        right = self.size_hint(left)
        self.guard_equals(left, sympy.Integer(right))
        return int(right)
   ```
   does that answer your question?

File: test/inductor/test_cpu_select_algorithm.py (hunk 2 above)
Comment (laithsakka) [reply test/inductor/test_cpu_select_algorithm.py]: that is True,  BMM Template will fail if Mdim, Kdim, Ndim are dynamic, but even with out this PR we probably might be able to came up with a program that fails it. ( if we use something else other than the reshape here no?)
I will hold of landing this until i get a green from you guys though

**Comment by laithsakka** [comment]
> > reshape_view_helper used to specialize unnecessarily.
> > and the new reshape does less specializations.
> 
> But is it supposed to override `torch._dynamo.mark_static`? Sorry, does `torch._dynamo.mark_static` mean that the corresponding dim shape should _always_ be specialized (so that the traced shape should always be used)? It isn't being used much in the codebase, so I'm not sure.

the code does not call torch._dynamo.mark_static on the Kdim, Ndim, Mdim it calls it on the input tensor , 
those variable are input to the reshape and are different happy to explain over VC on slack.

finally the XLA issue is unrelated we added sym_max to XLA , but i have a workaround in this PR that we will change once we update the XLA pin
https://github.com/pytorch/xla/pull/9291

**Comment by sanchitintel** [comment]
> evaluate_static_shape will guard on the hint meaning that a recompilation will happen if those change

Thanks for the info, @laithsakka!
This seems to suggest that the second change you suggested would help us trigger recompilation if `n` or `k` would be different at runtime (compared to the corresponding values traced at compile time).

```python
    if not has_free_unbacked_symbols(n):
        n = V.graph.sizevars.evaluate_static_shape(n)
    if not has_free_unbacked_symbols(k):
        k = V.graph.sizevars.evaluate_static_shape(k)
    if has_free_symbols((n, k)):
        return False
 ```

We'll try verifying it & will get back to you.

**Review by leslie-fang-intel — APPROVED** [review]

File: test/inductor/test_cpu_select_algorithm.py (hunk 2 above)
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_select_algorithm.py]: Thanks for the explanation!

**Comment by sanchitintel** [comment]
Hi @leslie-fang-intel, in its current form, this PR will cause a regression for the TorchBench workload corresponding to the UT that had been failing.

Did you approve this PR so that we may submit a PR of a patch after this PR would land?

Thanks!

**Comment by leslie-fang-intel** [comment]
> Hi @leslie-fang-intel, in its current form, this PR will cause a regression for the TorchBench workload corresponding to the UT that had been failing.
> 
> Did you approve this PR so that we may submit a PR of a patch after this PR would land?
> 
> Thanks!

Hi @sanchitintel, I didn't see the corresponding failure of TorchBench workload in the CI. Could you please post the link here?

**Comment by sanchitintel** [comment]
>  I didn't see the corresponding failure of TorchBench workload in the CI. Could you please post the link here?

@leslie-fang-intel, I'm guessing those workloads are run on a nightly or weekly basis?

This PR fixed the failing UT by forcing `n` and `k` to be static in the UT code.
However, the UT emulates a real workload.
When that workload would be run (after this PR would land), it wouldn't have static `n` and `k` because we wouldn't modify TorchBench models' source code. So, it'd cause a regression for the model(s) for which the UT had been added by @frost-intel.

@chuanqi129, IIRC, there are some weekly or nightly benchmarking CI jobs. Please confirm. Thanks!

**Comment by sanchitintel** [comment]
@laithsakka, please go ahead with landing the PR.

To unblock you, we would reactively add a patch if this PR would introduce a regression in our bi-weekly (I think fortnightly) benchmarking. Discussed with @leslie-fang-intel & it's even possible that the UT may not correspond to a real workload (although it was likely added because it corresponds to one), so we shouldn't keep you from landing this PR.

Sorry for the delay! Thank you for your patience!

**Comment by laithsakka** [comment]
> @laithsakka, please go ahead with landing the PR.
> 
> To unblock you, we would reactively add a patch if this PR would introduce a regression in our bi-weekly (I think fortnightly) benchmarking. Discussed with @leslie-fang-intel & it's even possible that the UT may not correspond to a real workload (although it was likely added because it corresponds to one), so we shouldn't keep you from landing this PR.
> 
> Sorry for the delay! Thank you for your patience!

Thank guys!

**Comment by laithsakka** [comment]
based on the discussion yesterday it sounds to me that my change here is wrong in the case of not unbacked. 
sounds to me that in the case of not unbacked we should actually run reshape_view_helper in case it returns strides that are different than those returned by the new simple version ... i do not think we have a guarantee the reshape_view_helper and compute_strides have the same results. 

I am thinking to revert and update the meta to call reshape_view_helper in case of unabcked free inputs.
now that will add back specializations that are not needed potentially but i guess we do not have a choice rn , unless we have a way to change eager reshape semantics.

**Comment by malfet** [comment]
Looks like this revert causes a trunk failures, may be because of some fixup changes that were landed after this one?
https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=aarch64&mergeEphemeralLF=true

**Comment by laithsakka** [comment]
> Looks like this revert causes a trunk failures, may be because of some fixup changes that were landed after this one? https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=aarch64&mergeEphemeralLF=true

shall i put a forward fix for those tests? I will take a look in a bit fixing them is not hard

File: test/inductor/test_cpu_select_algorithm.py (hunk 2 above)
Comment (laithsakka) [reply test/inductor/test_cpu_select_algorithm.py]: I am reverting this change i will change the new reshape to only be tirggered for unbacked so the specialization will happen now 
see my most recent comment for the reasoning. 
@leslie-fang-intel  @sanchitintel
